Client/src/hooks/dataRestriction.ts

```diff
@@ -16,14 +19,148 @@ import {
 const STUDY_ACTION_CLASS_NAME = 'study-action';
 
 const STUDY_ID_DATA_ATTRIBUTE = 'data-study-id';
 const ARGS_DATA_ATTRIBUTE = 'data-args';
 
+export type CompleteApprovalStatus = 'approved' | 'not-approved' | 'study-not-found';
+
+export type ApprovalStatus = CompleteApprovalStatus | 'loading';
+
+/**
+ * @param studyId
+ * @param action
+ * @returns An ApprovalStatus signifying if the user is permitted to
+ * perform the specified "action" for the study with id "studyId"
+ */
+export function useApprovalStatus(studyId: string, action: string) {
+  const [approvalStatus, setApprovalStatus] = useState<ApprovalStatus>(
+    'loading'
+  );
+
+  const attemptActionTask = useAttemptActionTask(studyId, action);
+
+  useEffect(() => {
+    setApprovalStatus('loading');
+
+    return attemptActionTask.run(setApprovalStatus);
+  }, [attemptActionTask]);
+
+  return approvalStatus;
+}
+
+/**
+ * @param studyId
+ * @param action
+ * @param onApprovalStatusCheckComplete
+ * @returns A callback, which, when executed:
+ *
+ * (1) attempts to perform "action" for the study with "studyId"
+ *
+ * (2) executes "onApprovalStatusCheckComplete" once the user's permission to
+ * perform the action has been checked
+ *
+ * NOTE: whenever the passed "studyId", "action", or "onApprovalStatusCheckComplete"
+ * arguments have changed, or the component which invoked this hook has been unmounted,
+ * all pending executions of "onApprovalStatusCheckComplete" will be canceled
+ */
+export function useScopedAttemptActionCallback(
+  studyId: string,
+  action: string,
+  onApprovalStatusCheckComplete: (
+    approvalStatus: CompleteApprovalStatus
+  ) => void
+) {
+  const attemptActionTask = useAttemptActionTask(studyId, action);
+
+  const cancelCbs = useRef<Set<() => void>>(new Set());
+
+  const attemptActionCallback = useCallback(() => {
+    const cancelCb = attemptActionTask.run(onApprovalStatusCheckComplete);
+    cancelCbs.current.add(cancelCb);
+  }, [attemptActionTask, onApprovalStatusCheckComplete]);
+
+  useEffect(
+    () => () => {
+      for (const cancelCb of cancelCbs.current) {
+        cancelCb();
+      }
+
+      cancelCbs.current.clear();
+    },
+    [attemptActionCallback]
+  );
+
+  return attemptActionCallback;
+}
+
+/**
+ * @param studyId
+ * @param action
+ * @returns A Task which, when run, will:
+ * (1) check if the user is permitted to perform "action" for the study with id "studyId"
+ * (2) if appropriate, open a Data Restriction modal
+ * (3) resolve to the appropriate ApprovalStatus
+ */
+export function useAttemptActionTask<E>(
+  studyId: string,
+  action: string
+): Task<CompleteApprovalStatus, E> {
+  const dispatch = useDispatch();
+  const wdkDependencies = useContext(WdkDepdendenciesContext);
+
+  if (wdkDependencies == null) {
+    throw new Error(
+      'In order to "useAttemptActionTask", WdkDependenciesContext must be provided.'
+    );
+  }
+
+  return useMemo(
+    () =>
+      new Task<any, E>(function (fulfill, reject) {
```

Comment:
Does `Task<CompleteApprovalStatus, E>` not work here?

Reply:
Good question! `Task<CompleteApprovalStatus, E>` won't work because this intermediate task is actually a `Task<DataRestrictionReduxAction, E>`, where `DataRestrictionReduxAction` is one of the three types of redux actions defined [here](https://github.com/VEuPathDB/EbrcWebsiteCommon/blob/5bf96666ce6e6d824592d0cf914363cdf45a40e5/Client/src/App/DataRestriction/DataRestrictionActionCreators.js#L29).

The reason for the `any` is that the contents of `DataRestrictionActionCreators` has not been TypeScript-ified. Why don't I try doing that and get rid of the `any` here? (In doing so, I may use, in the short term, a few `any`s, especially for the return type of `fetchStudies`.)